modxcms/revolution
# MODX 3 refactoring

### What does it do?
- PSR-4 autoloading
- Core model updated to xPDO 3.0
- Refactored processors
- Removed modXMLRPCResource and modJSONRPCResource (and also their Response classes)
- Replaced MagpieRSS to SimplePie
- Core files formatted to PSR-2
- Updated build, setup, model generator, tests, etc. Everything needed to make it work.

### Why is it needed?
To make MODX modern again.

### Related issue(s)/PR(s)
#11272 

### How to test
1. `composer create-project modx/revolution www 3.x-dev` where `www` is the directory to install.
When it will ask `Do you want to remove the existing VCS (.git, .svn..) history? [Y,n]?` answer **n**.
2. Fetch this PR: `cd www && git fetch origin pull/13900/head:3x-refactoring`.
3. Checkout to the new branch: `git checkout 3x-refactoring`
4. Update composer: `composer update`
5. Build core files: `php ./_build/transport.core.php`
6. Open your website in browser and run Setup.
7. Login to manager and disable system setting `compress_js`. Clear browser cache and after this you will see no more errors about wrong processors paths.

### Compatibility
Sorry, but after moving MODX to namespace, we will need to update all complicated extras to this modern codebase. 

Of course, there are some compatibility workarounds like `MODX::loadClass()` method and automatic aliases to old class names, but there can't be 100% support of old codebase.

For now it was tested with:
- [x] Ace
- [x] pdoTools
- [x] Theme.Bootstrap
- [x] MinifyX
- [ ] ms2Gallery - installs, but can't work with `modMediaSource` to generate previews
- [ ] miniShop2 - installs, but can't extend `modResource`
- [ ] mSearch2 - installs, but cant't index resources.
- [ ] Office - installs and can authentificate users, but can't register them, because of extending `modUserCreateProcessor`

I think that `modResource` extending packages will require more work (Articles, Collections, Tickets, miniShop2, etc). Because now they should extend `MODX\modResource`, but not `modResource` and it will require to change their model to version 3.0
 
### Whats next?
I didn't finished yet, but now I want to show and discuss this changes. Then I want to move further:
- Remove all `getService()` calls from the core
- Rewrite all `getObject('modResource')` to `getObject('MODX\modResource')` because it will not require to get correct class name with `MODX::loadClass()`.
- ? 
- Update all necessary extras.
